flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoder.java

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * Decodes messages from the fragmentary netty buffers. This decoder assumes the
+ * messages have the following format:
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ *
+ * This decoder decodes the frame header and delegates the other following work
+ * to corresponding message parsers according to the message type. During the process
+ * of decoding, the decoder and parsers try best to eliminate copying. For the frame
+ * header and message header, it only cumulates data when they span multiple input buffers.
+ * For the buffer part, it copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class NettyMessageClientDecoder extends ChannelInboundHandlerAdapter {
+
+	/** The message parser for buffer response. */
+    private final NettyMessageDecoderDelegate bufferResponseDecoderDelegate;
+
+    /** The message parser for other messages other than buffer response. */
+	private final NettyMessageDecoderDelegate nonBufferResponseDecoderDelegate;
+
+	/** The cumulation buffer for the frame header part. */
+	private ByteBuf frameHeaderBuffer;
+
+	/**
+	 * The chosen message parser for the current message. If it is null, then
+	 * we are decoding the frame header part, otherwise we are decoding the actual
+	 * message.
+	 */
+	private NettyMessageDecoderDelegate currentDecoderDelegate;
+
+    NettyMessageClientDecoder(NetworkClientHandler networkClientHandler) {
+        this.bufferResponseDecoderDelegate = new BufferResponseDecoderDelegate(networkClientHandler);
+        this.nonBufferResponseDecoderDelegate = new NonBufferResponseDecoderDelegate();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        bufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+        nonBufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+
+		frameHeaderBuffer = ctx.alloc().directBuffer(FRAME_HEADER_LENGTH);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+		bufferResponseDecoderDelegate.release();
+		nonBufferResponseDecoderDelegate.release();
+
+		frameHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.isReadable()) {
+				if (currentDecoderDelegate == null) {
```

Comment:
It maybe better to adjust the logic a bit, to judge the hot path firstly.

```
while (data.isReadable()) {
          if (currentDecoderDelegate != null) {
	          NettyMessageDecoderDelegate.ParseResult result =currentDecoderDelegate.onChannelRead(data);
		 if (!result.finished) {
			return;
		}
		if (result.message != null) {
			ctx.fireChannelRead(result.message);
		}
		currentDecoderDelegate = null;
		frameHeaderBuffer.clear();
        }
        decodeFrameHeader(data)
}
```

Reply:
Have moved the logic of decoding frame header into separate method. 

But I'm slightly tends to keep the order since it should be easier to understand if the branch is consistent with the message structure.

Reply:
If we refer to `StreamTaskNetworkInput#emitNext`, it always judges the `currentRecordDeserializer != null` at the front which is the same case with the decoder. I do not think it is very hard to understand the logic.

From another aspect it is better to give the frequent hot path firstly. Another concern was that we can avoid judging the condition of `data.isReadable()` twice which seems a bit redundant.

Reply:
Agree with we could simplify the branches here. Have modified accordingly.